File: node/src/inmemorycheckpointer.ts
/*
 * Copyright 2022 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { Checkpointer } from './checkpointer';
import { ChaincodeEvent } from './chaincodeevent';

/**
 * In-memory checkpointer class used to persist checkpointer state in memory.
 */

export class InMemoryCheckPointer implements Checkpointer {

	#blockNumber?: bigint;
	#transactionID?: string;

	checkpointBlock(blockNumber: bigint): void {
		this.#blockNumber = blockNumber + BigInt(1);
		this.#transactionID = undefined;
	}

	checkpointTransaction(blockNumber: bigint, transactionId: string): void {
		this.#blockNumber = blockNumber;
		this.#transactionID = transactionId;
	}

	checkpointChaincodeEvent(event: ChaincodeEvent): void {
		this.checkpointTransaction(event.blockNumber,event.transactionId);
	}

	getBlockNumber(): bigint | undefined {
		return this.#blockNumber;
	}

	getTransactionId(): string | undefined {
		return this.#transactionID;
	}
}

File: pkg/client/inmemorycheckpointer.go
/*
Copyright 2022 IBM All Rights Reserved.

SPDX-License-Identifier: Apache-2.0
*/

package client

type InMemoryCheckpointer struct {
	blockNumber   uint64
	transactionID string
}

func (c *InMemoryCheckpointer) CheckpointBlock(blockNumber uint64) error {
	c.blockNumber = blockNumber + 1
	c.transactionID = ""
	return nil
}

func (c *InMemoryCheckpointer) CheckpointTransaction(blockNumber uint64, transactionID string) error {
	c.blockNumber = blockNumber
	c.transactionID = transactionID
	return nil
}

func (c *InMemoryCheckpointer) CheckpointChaincodeEvent(event *ChaincodeEvent) error {
	err := c.CheckpointTransaction(event.BlockNumber, event.TransactionID)
	return err
}

func (c *InMemoryCheckpointer) BlockNumber() uint64 {
	return c.blockNumber
}

func (c *InMemoryCheckpointer) TransactionID() string {
	return c.transactionID
}
